overlord/assertstate/assertstate.go

```diff
@@ -887,10 +891,97 @@ func TryEnforceValidationSets(st *state.State, validationSets []string, userID i
 	return addCurrentTrackingToValidationSetsHistory(st)
 }
 
-// EnforceValidationSet tries to fetch the given validation set and enforce it.
+// EnforceValidationSets takes a map from validation set specifiers in the
+// foo/bar=N syntax to ValidationSet assertions and enforces the sets, if possible.
+// It doesn't fetch any validation sets.
+func EnforceValidationSets(st *state.State, vsMap map[string]*asserts.ValidationSet, snaps []*snapasserts.InstalledSnap, ignoreValidation map[string]bool, userID int) error {
+	user, err := userFromUserID(st, userID)
+	if err != nil {
+		return err
+	}
+
+	sto := snapstate.Store(st, nil)
+	retrieve := func(ref *asserts.Ref) (asserts.Assertion, error) {
+		return sto.Assertion(ref.Type, ref.PrimaryKey, user)
+	}
+
+	db := cachedDB(st)
+	batch := asserts.NewBatch(nil)
```

Comment:
this code could use the doFetch helper by moving the for loop below inside the fetching I think. We might have to tweak doFetch not do the committing though or do it optionally